kura/org.eclipse.kura.nm/src/main/java/org/eclipse/kura/nm/NMDbusConnector.java

```diff
@@ -274,53 +274,64 @@ public synchronized void apply() throws DBusException {
             logger.warn("No cached network configuration found.");
             return;
         }
-        doApply(this.cachedConfiguration);
+        doApply(Optional.empty(), this.cachedConfiguration);
+    }
+
+    public synchronized void apply(String deviceId) throws DBusException {
+        if (Objects.isNull(this.cachedConfiguration)) {
+            logger.warn("No cached network configuration found.");
+            return;
+        }
+        if (Objects.nonNull(deviceId) && !deviceId.isEmpty()) {
+            doApply(Optional.of(deviceId), this.cachedConfiguration);
+        } else {
+            doApply(Optional.empty(), this.cachedConfiguration);
+        }
     }
 
-    private synchronized void doApply(Map<String, Object> networkConfiguration) throws DBusException {
+    private synchronized void doApply(Optional<String> deviceIdToBeConfigured, Map<String, Object> networkConfiguration)
+            throws DBusException {
         logger.info("Applying configuration using NetworkManager Dbus connector");
         try {
             configurationEnforcementDisable();
 
             NetworkProperties properties = new NetworkProperties(networkConfiguration);
-
-            List<String> configuredInterfaces = properties.getStringList("net.interfaces");
-            manageConfiguredInterfaces(configuredInterfaces, properties);
-
-            List<Device> availableInterfaces = getAllDevices();
-            manageNonConfiguredInterfaces(configuredInterfaces, availableInterfaces);
+            List<String> configuredInterfaceIds = properties.getStringList("net.interfaces");
+
+            List<Device> availableDevices = getAllDevices();
+            availableDevices.forEach(device -> {
+                String deviceId = null;
+                try {
+                    deviceId = getDeviceId(device.getObjectPath());
+                } catch (DBusException e) {
+                    logger.error("Unable to retrieve the device Id for device path {}", device.getObjectPath(), e);
+                }
+                if (shouldBeConfigured(deviceId, deviceIdToBeConfigured)) {
+                    try {
+                        if (configuredInterfaceIds.contains(deviceId)) {
+                            manageConfiguredInterface(device, deviceId, properties);
+                        } else {
+                            manageNonConfiguredInterface(device, deviceId);
+                        }
+                    } catch (DBusException | DBusExecutionException | IllegalArgumentException
+                            | NoSuchElementException e) {
+                        logger.error("Unable to configure device {}, skipping", deviceId, e);
+                    }
+                }
+            });
         } finally {
             configurationEnforcementEnable();
         }
     }
```

Comment:
This is way too unnecessarily complicated. I propose the following:

```suggestion
            List<String> configuredInterfaceIds = properties.getStringList("net.interfaces");

            List<Device> availableDevices = getAllDevices();
            availableDevices.forEach(device -> {
                String deviceId = getDeviceId(device.getObjectPath());
                if (shouldBeConfigured(deviceId, deviceIdToBeConfigured)) {
                    if (configuredInterfaceIds.contains(deviceId)) {
                        manageConfiguredInterface(device, deviceId, properties);
                    } else {
                        manageNonConfiguredInterface(device, deviceId);
                    }
                }
            });
        } catch (DBusException | DBusExecutionException | IllegalArgumentException
                | NoSuchElementException e) {
            logger.error("Unable to configure device {}, skipping", deviceId, e); // If you don't have the deviceId you can use the DBus path here
        } finally {
            configurationEnforcementEnable();
        }
    }
```